File: Data Structures & Algorithms/delete-node-in-a-bst/submission-2.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def find_successor(self, root: Optional[TreeNode]) -> Optional[TreeNode]:
        while root and root.left:
            root = root.left
        return root

    def deleteNode(self, root: Optional[TreeNode], key: int) -> Optional[TreeNode]:
        if root is None:
            return None
        if key < root.val:
            root.left = self.deleteNode(root.left, key)
        elif key > root.val:
            root.right = self.deleteNode(root.right, key)
        else:
            if root.left is None and root.right is None:
                root = None
                return None
            elif root.left is None or root.right is None:
                temp: Optional[TreeNode] = root.left if (root.left) else root.right
                root = None
                return temp
            else :
                temp : Optional[TreeNode] = self.find_successor(root.right)
                root.val = temp.val
                root.right = self.deleteNode(root.right, temp.val)
        return root

File: Data Structures & Algorithms/binary-tree-from-preorder-and-inorder-traversal/submission-2.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    unordered_map<int,int> mp;
    TreeNode* buildTree(vector<int>& pre, vector<int>& in) {
        if(pre.size() == 1){
            TreeNode* root = new TreeNode(pre[0]);
            return root;
        }
        mp.clear();
        for(int i=0;i<in.size();i++){
            mp[in[i]] = i ;
        }
        int start = 0;
        return helper(pre, start, 0, in.size()-1);
    }
    TreeNode* helper(vector<int>& pre, int& start, int l, int r){
        if(l > r)return nullptr;
        int val = pre[start++];
        TreeNode* root = new TreeNode(val);
        if(l == r)return root;
        int index = mp[val];
        root->left = helper(pre, start, l, index-1);
        root->right = helper(pre, start, index+1, r);
        return root;
    }
};
